### ISSUE TITLE
Probit regression fails to converge when it should

### ISSUE THREAD

**Issue by lendle** [JuliaStats/GLM.jl#99]
From [this discussion](https://groups.google.com/forum/#!topic/julia-stats/8ngcUHpcDxQ).

``` julia
julia> using DataFrames, GLM

julia> srand(10001)

julia> N=10000
10000

julia> distX = Normal(0,3)
Normal(μ=0.0, σ=3.0)

julia> X = rand(distX,N);

julia> ɛ = randn(N);

julia> Y = X.>ɛ;

julia> df = DataFrame(X=[X],Y=[Y]);

julia> fittedmodel = glm(Y ~ X, df, Binomial(), ProbitLink(), verbose=true);
1: 5082.253852706774, 0.25854214488487215
2: 4555.984395197572, 0.11551169008917996
3: NaN, NaN
<...snip...>
30: NaN, NaN
WARNING: failure to converge in 30 iterations

julia> fittedmodel.model.pp.beta0
2-element Array{Float64,1}:
 NaN
 NaN
```

Here's a plot of the data set:
![plot](https://i.imgur.com/SLfTr7y.png)
There's plenty of overlap near X=0, so the issue is not complete separation. Fitting converges without issue with the `LogitLink()`.

R converges with an estimate close to 1 for the coefficient of X, and a reasonable SE:

``` R

> n=10000
> X=rnorm(n,0,3)
> eps=rnorm(n)
> Y = X > eps
> fit = glm(Y~X, family=binomial(link="probit"))
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
> summary(fit)

Call:
glm(formula = Y ~ X, family = binomial(link = "probit"))

Deviance Residuals:
    Min       1Q   Median       3Q      Max
-3.3259  -0.2189   0.0000   0.2104   3.3539

Coefficients:
            Estimate Std. Error z value Pr(>|z|)
(Intercept) 0.004144   0.021013   0.197    0.844
X           0.998480   0.020138  49.582   <2e-16 ***

---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 13862.9  on 9999  degrees of freedom
Residual deviance:  4561.8  on 9998  degrees of freedom
AIC: 4565.8

Number of Fisher Scoring iterations: 7
```

**Comment by lendle** [comment]
The link, inverse link, and mueta functions for ProbitLink() are correct I think so I am not sure where the trouble is coming from.

**Comment by andreasnoack** [comment]
I think I have a fix for this. The precision can be improved in the inverse probit functions, in the calculation of the the Bernoulli variance and in the residual deviance calculation. With these changes, you example converges, but unfortunately they cause a test fail in the test in line 15. I'm not sure why.

**Comment by andreasnoack** [comment]
This is not completely fixed by #100 although the example above now works. The problem is only postponed until the separation becomes more severe. In the estimation weight calculation, there is a `0/0` term causing `NaN`s when observation are far away from the mean. I think it is safe to set these to `0` instead of `NaN`, but we probably want to print a warning when that happens.

**Comment by andreasnoack** [comment]
Let's close this issue. The example I shared in the previous post is a perfect separation case so this is clearly different from the original case.
